### ISSUE TITLE
[Urgent] Wrong dynamic URL being delivered 

### ISSUE THREAD

**Issue by trafiko** [mautic/mautic#4155]
What type of report is this:

| Q  | A
| ---| ---
| Bug report? | Y
| Feature request? | N 
| Enhancement? | N

## Description:
Two of our contact fields are URLs. We sent out emails including {leadfield=urlfieldname} inside linked text or buttons. Contacts are getting every dynamic field correctly, except the two URLs. They are getting URLs from other contacts in the database, which is obviously a total disaster. List size is 3000+. ¿What could of happened? ¿How do we solve this issue?

## If a bug:

| Q   | A
| --- | ---
| Mautic version | v2.2.1
| PHP version | 5.6.26

### Steps to reproduce:
1. Create a field for a URL for your contacts.
2. Input a different URL for each contact.
3. Send Email with a button linking to this URLs
4. Users get URLs related to different contacts. 
 
### Log errors: 

_Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable._

**Comment by panchicore** [comment]
Logs we have:
mautic_prod-2017-05-16.php  mautic_prod-2017-05-20.php  mautic_prod-2017-05-23.php  mautic_prod-2017-05-25.php
mautic_prod-2017-05-17.php  mautic_prod-2017-05-22.php  mautic_prod-2017-05-24.php

```
# more mautic_prod-2017-05-* |grep "mautic.ERROR"
[2017-05-16 21:46:51] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /a2billing/common/javascript/misc.js" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /a2billing/common/javascript/misc.js at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-16 21:48:47] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /a2billing/common/javascript/misc.js" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /a2billing/common/javascript/misc.js at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-17 03:47:46] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /admin/images/tango.png" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /admin/images/tango.png at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-20 06:11:13] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /admin/config.php" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /admin/config.php at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-22 14:24:00] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /recordings" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /recordings at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-23 13:37:39] mautic.ERROR: An unexpected 500 code was returned while attempting to fetch updates.  The message received was: <!DOCTYPE html> <html lang="en">   <head>     <meta charset="utf-8">     <!--<meta http-equiv="X-UA-Compatible" content="IE=edge"><script type="text/javascript">window.NREUM||(NREUM={}),__nr_require=function(e,n,t){function r(t){if(!n[t]){var o=n[t]={exports:{}};e[t][0].call(o.exports,function(n){var o=e[t][1][n];return r(o||n)},o,o.exports)}return n[t].exports}if("function"==typeof __nr_require)return __nr_require;for(var o=0;o<t.length;o++)r(t[o]);return r}({1:[function(e,n,t){function r(){}function o(e,n,t){return function(){return i(e,[c.now()].concat(u(arguments)),n?null:this,t),n?void 0:this}}var i=e("handle"),a=e(2),u=e(3),f=e("ee").get("tracer"),c=e("loader"),s=NREUM;"undefined"==typeof window.newrelic&&(newrelic=s);var p=["setPageViewName","setCustomAttribute","setErrorHandler","finished","addToTrace","inlineHit","addRelease"],d="api-",l=d+"ixn-";a(p,function(e,n){s[n]=o(d+n,!0,"api")}),s.addPageAction=o(d+"addPageAction",!0),s.setCurrentRouteName=o(d+"routeName",!0),n.exports=newrelic,s.interaction=function(){return(new r).get()};var m=r.prototype={createTracer:function(e,n){var t={},r=this,o="function"==typeof n;return i(l+"tracer",[c.now(),e,t],r),function(){if(f.emit((o?"":"no-")+"fn-start",[c.now(),r,o],t),o)try{return n.apply(this,arguments)}finally{f.emit("fn-end",[c.now()],t)}}}};a("setName,setAttribute,save,ignore,onEnd,getContext,end,get".split(","),function(e,n){m[n]=o(l+n)}),newrelic.noticeError=function(e){"string"==typeof e&&(e=new Error(e)),i("err",[e,c.now()])}},{}],2:[function(e,n,t){function r(e,n){var t=[],r="",i=0;for(r in e)o.call(e,r)&&(t[i]=n(r,e[r]),i+=1);return t}var o=Object.prototype.hasOwnProperty;n.exports=r},{}],3:[function(e,n,t){function r(e,n,t){n||(n=0),"undefined"==typeof t&&(t=e?e.length:0);for(var r=-1,o=t-n||0,i=Array(o<0?0:o);++r<o;)i[r]=e[n+r];return i}n.exports=r},{}],4:[function(e,n,t){n.exports={exists:"undefined"!=typeof window.performance&&window.performance.timing&&"undefined"!=typeof window.performance.timing.navigationStart}},{}],ee:[function(e,n,t){function r(){}function o(e){function n(e){return e&&e instanceof r?e:e?f(e,u,i):i()}function t(t,r,o,i){if(!d.aborted||i){e&&e(t,r,o);for(var a=n(o),u=m(t),f=u.length,c=0;c<f;c++)u[c].apply(a,r);var p=s[y[t]];return p&&p.push([b,t,r,a]),a}}function l(e,n){v[e]=m(e).concat(n)}function m(e){return v[e]||[]}function w(e){return p[e]=p[e]||o(t)}function g(e,n){c(e,function(e,t){n=n||"feature",y[t]=n,n in s||(s[n]=[])})}var v={},y={},b={on:l,emit:t,get:w,listeners:m,context:n,buffer:g,abort:a,aborted:!1};return b}function i(){return new r}function a(){(s.api||s.feature)&&(d.aborted=!0,s=d.backlog={})}var u="nr@context",f=e("gos"),c=e(2),s={},p={},d=n.exports=o();d.backlog=s},{}],gos:[function(e,n,t){function r(e,n,t){if(o.call(e,n))return e[n];var r=t();if(Object.defineProperty&&Object.keys)try{return Object.defineProperty(e,n,{value:r,writable:!0,enumerable:!1}),r}catch(i){}return e[n]=r,r}var o=Object.prototype.hasOwnProperty;n.exports=r},{}],handle:[function(e,n,t){function r(e,n,t,r){o.buffer([e],r),o.emit(e,n,t)}var o=e("ee").get("handle");n.exports=r,r.ee=o},{}],id:[function(e,n,t){function r(e){var n=typeof e;return!e||"object"!==n&&"function"!==n?-1:e===window?0:a(e,i,function(){return o++})}var o=1,i="nr@id",a=e("gos");n.exports=r},{}],loader:[function(e,n,t){function r(){if(!x++){var e=h.info=NREUM.info,n=d.getElementsByTagName("script")[0];if(setTimeout(s.abort,3e4),!(e&&e.licenseKey&&e.applicationID&&n))return s.abort();c(y,function(n,t){e[n]||(e[n]=t)}),f("mark",["onload",a()+h.offset],null,"api");var t=d.createElement("script");t.src="https://"+e.agent,n.parentNode.insertBefore(t,n)}}function o(){"complete"===d.readyState&&i()}function i(){f("mark",["domContent",a()+h.offset],null,"api")}function a(){return E.exists&&performance.now?Math.round(performance.now()):(u=Math.max((new Date).getTime(),u))-h.offset}var u=(new Date).getTime(),f=e("handle"),c=e(2),s=e("ee"),p=window,d=p.document,l="addEventListener",m="attachEvent",w=p.XMLHttpRequest,g=w&&w.prototype;NREUM.o={ST:setTimeout,CT:clearTimeout,XHR:w,REQ:p.Request,EV:p.Event,PR:p.Promise,MO:p.MutationObserver};var v=""+location,y={beacon:"bam.nr-data.net",errorBeacon:"bam.nr-data.net",agent:"js-agent.newrelic.com/nr-1026.min.js"},b=w&&g&&g[l]&&!/CriOS/.test(navigator.userAgent),h=n.exports={offset:u,now:a,origin:v,features:{},xhrWrappable:b};e(1),d[l]?(d[l]("DOMContentLoaded",i,!1),p[l]("load",r,!1)):(d[m]("onreadystatechange",o),p[m]("onload",r)),f("mark",["firstbyte",u],null,"api");var x=0,E=e(4)},{}]},{},["loader"]);</script>-->     <meta name="viewport" content="width=device-width, initial-scale=1">     <meta name="description" content="">     <meta name="author" content="">     <link rel="shortcut icon" href="templates/mautic/favicon.ico">         <link href='//fonts.googleapis.com/css?family=Lato:300,400,300italic,400italic' rel='stylesheet' type='text/css'>     <link href='//fonts.googleapis.com/css?family=Montserrat:400,700' rel='stylesheet' type='text/css'>         <!-- Bootstrap core CSS -->     <link href="//maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">      <!-- Resources -->     <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.1.0/css/font-awesome.min.css" rel="stylesheet">      <!-- Custom styles for this template -->     <link href="templates/mautic/css/style.css" rel="stylesheet">     <link href="templates/mautic/css/error.css" rel="stylesheet">      <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->     <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->     <!--[if lt IE 9]>       <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>       <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>     <![endif]--> 	</head> 	<body> 		<div class="container"> 			<div class="row"> 				<div class="col-xs-6"> 								<h1>Oh Dear!</h1> 							<h3>Well this is awkward. Clearly you came here expecting something, but it's not here. We're sorry if we told you it was. May I suggest a hasty return to the home page?</h3>							<a href="/" class="btn btn-cta btn-cta-primary">Home</a> 				</div> 				<div class="col-xs-6"> 					<img src="http://www.mautic.org/media/images/mautibot_mouth_o.png" width="250px" /> 				</div> 		</div> 	<script type="text/javascript">window.NREUM||(NREUM={});NREUM.info={"beacon":"bam.nr-data.net","licenseKey":"dff2df6742","applicationID":"10952386","transactionName":"Zl0EZ0FWDRJUURIKC18XM0FaGAoPUVceTRRZSA==","queueTime":0,"applicationTime":12,"atts":"ShoHEQlMHhw=","errorBeacon":"bam.nr-data.net","agent":""}</script></body> </html>  [] []
[2017-05-24 05:01:19] mautic.ERROR: Error: Call to a member function isRendered() on null - in file /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php - at line 215 [] []
[2017-05-24 15:45:11] mautic.ERROR: Error: Call to a member function isRendered() on null - in file /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php - at line 215 [] []
[2017-05-24 15:45:18] mautic.ERROR: Error: Call to a member function isRendered() on null - in file /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php - at line 215 [] []
[2017-05-24 15:45:29] mautic.ERROR: Error: Call to a member function isRendered() on null - in file /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php - at line 215 [] []
[2017-05-24 15:45:51] mautic.ERROR: Error: Call to a member function isRendered() on null - in file /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php - at line 215 [] []
[2017-05-24 15:45:59] mautic.ERROR: Error: Call to a member function isRendered() on null - in file /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php - at line 215 [] []
[2017-05-24 17:57:25] mautic.ERROR: [MAIL ERROR] Connection to tcp://smtp.sendgrid.net:587 Timed Out Log data: ++ Starting Mautic\EmailBundle\Swiftmailer\Transport\SendgridTransport !! Connection to tcp://smtp.sendgrid.net:587 Timed Out (code: 0) [] []
[2017-05-24 17:58:10] mautic.ERROR: [MAIL ERROR] Connection to tcp://smtp.sendgrid.net:587 Timed Out Log data: ++ Starting Mautic\EmailBundle\Swiftmailer\Transport\SendgridTransport << 220 SG ESMTP service ready at ismtpd0005p1iad1.sendgrid.net  >> EHLO mailing.URLREPLACEDHERE.com  << 250-smtp.sendgrid.net 250-8BITMIME 250-PIPELINING 250-SIZE 31457280 250-STARTTLS 250-AUTH PLAIN LOGIN 250 AUTH=PLAIN LOGIN  !! Connection to tcp://smtp.sendgrid.net:587 Timed Out (code: 0) Log data: ++ Starting Mautic\EmailBundle\Swiftmailer\Transport\SendgridTransport << 220 SG ESMTP service ready at ismtpd0005p1iad1.sendgrid.net  >> EHLO mailing.URLREPLACEDHERE.com  << 250-smtp.sendgrid.net 250-8BITMIME 250-PIPELINING 250-SIZE 31457280 250-STARTTLS 250-AUTH PLAIN LOGIN 250 AUTH=PLAIN LOGIN  !! Connection to tcp://smtp.sendgrid.net:587 Timed Out (code: 0) !! Connection to tcp://smtp.sendgrid.net:587 Timed Out Log data: ++ Starting Mautic\EmailBundle\Swiftmailer\Transport\SendgridTransport << 220 SG ESMTP service ready at ismtpd0005p1iad1.sendgrid.net  >> EHLO mailing.URLREPLACEDHERE.com  << 250-smtp.sendgrid.net 250-8BITMIME 250-PIPELINING 250-SIZE 31457280 250-STARTTLS 250-AUTH PLAIN LOGIN 250 AUTH=PLAIN LOGIN  !! Connection to tcp://smtp.sendgrid.net:587 Timed Out (code: 0) (code: 0) [] []
[2017-05-24 17:58:40] mautic.ERROR: [MAIL ERROR] Connection could not be established with host smtp.sendgrid.net [Connection timed out #110] Log data: ++ Starting Mautic\EmailBundle\Swiftmailer\Transport\SendgridTransport !! Connection could not be established with host smtp.sendgrid.net [Connection timed out #110] (code: 0) [] []
[2017-05-25 08:19:15] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 08:19:22] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 09:30:09] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 09:30:10] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 12:05:47] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 12:05:48] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 13:02:50] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 13:02:51] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 13:03:24] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []
[2017-05-25 13:03:25] mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "URL solicitada no encontrada: /email/" at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php line 465 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): URL solicitada no encontrada: /email/ at /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php:465)"} []

```
mautic.WARNING just many of those:

```
[2017-05-16 21:46:51] mautic.WARNING: IP LOOKUP: The file "/var/www/html/app/cache/prod/../ip_data/GeoLite2-City.mmdb" does not exist or is not readable. [] []
```

Apache ssl_error_logs

```
==> ssl_error_log <==
[Wed May 24 05:01:19 2017] [error] [client 217.118.241.247] PHP Fatal error:  Call to a member function isRendered() on null in /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php on line 215, referer: https://mailing.URLREPLACEDHERE.com/s/contacts/view/7821
[Wed May 24 15:45:11 2017] [error] [client 186.147.182.138] PHP Fatal error:  Call to a member function isRendered() on null in /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php on line 215, referer: https://mailing.URLREPLACEDHERE.com/s/contacts/view/19413
[Wed May 24 15:45:18 2017] [error] [client 186.147.182.138] PHP Fatal error:  Call to a member function isRendered() on null in /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php on line 215, referer: https://mailing.URLREPLACEDHERE.com/s/contacts/view/19413
[Wed May 24 15:45:29 2017] [error] [client 186.147.182.138] PHP Fatal error:  Call to a member function isRendered() on null in /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php on line 215, referer: https://mailing.URLREPLACEDHERE.com/s/contacts/view/19413
[Wed May 24 15:45:51 2017] [error] [client 186.147.182.138] PHP Fatal error:  Call to a member function isRendered() on null in /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php on line 215, referer: https://mailing.URLREPLACEDHERE.com/s/contacts/view/19413
[Wed May 24 15:45:59 2017] [error] [client 186.147.182.138] PHP Fatal error:  Call to a member function isRendered() on null in /var/www/html/app/bundles/LeadBundle/Views/Lead/form.html.php on line 215, referer: https://mailing.URLREPLACEDHERE.com/s/contacts/view/19413
```

**Comment by escopecz** [comment]
This has been fixed in Mautic 2.8.1 (https://github.com/mautic/mautic/pull/3766). Next time please upgrade to the latest version before submitting a bug report because there is a chance it has been fixed already.

**Comment by panchicore** [comment]
@escopecz update causes an error https://github.com/mautic/mautic/issues/4156 thank you in advance.
